### ISSUE TITLE
Inconsistent warnings regarding interpolation-only syntax on the left side of assignment

### ISSUE THREAD

**Issue by aaronsteers** [hashicorp/terraform#27037]
### Terraform Version

```
Terraform v0.14.0-rc1
+ provider registry.terraform.io/hashicorp/archive v2.0.0
+ provider registry.terraform.io/hashicorp/aws v3.17.0
+ provider registry.terraform.io/hashicorp/http v2.0.0
+ provider registry.terraform.io/hashicorp/local v2.0.0
+ provider registry.terraform.io/hashicorp/null v3.0.0
+ provider registry.terraform.io/hashicorp/random v3.0.0
```

### Terraform Configuration Files

```terraform
locals {
  dynamic     = "key_b"
  my_map_json = jsonencode({
    key_a              = "something"
    "${local.dynamic}" = "else"
  })
  # ...
}
```

### Expected Behavior

This _should not_ emit a warning regarding interpolation-only syntax. 

If interpolation is removed, the following error is displayed:

```
Error: Ambiguous attribute key

  on ../../catalog/aws/ml-ops/main.tf line 272, in module "my_module_name":
 272:             local.dynamic = "else"

If this expression is intended to be a reference, wrap it in parentheses. If
it's instead intended as a literal name containing periods, wrap it in quotes
to create a string literal.
```

### Actual Behavior

```
Warning: Interpolation-only expressions are deprecated

  on main.tf line 391, in module "my_module_name":
 391:               "${local.dynamic}" = "else"

Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.

...
```

### Steps to Reproduce

1. Use dynamic key names with interpolation syntax.
2. Run any terraform operation and receive the warning about interpolation-only syntax being deprecated.
3. If you follow the advice, you receive a new error (as described above) and must revert.
4. The warning persists indefinitely.

### Additional Context

The becomes more of an issue when applying the new opinionated experience in 0.14.0. I have not tested and confirm that this is auto-formatted but there's a risk that it might be in future versions (if not already).

**Comment by alisdair** [comment]
Hi @aaronsteers, thanks for reporting this.

We actually just improved this very warning recently in #26949. The output is now more useful:

```hcl
Warning: Interpolation-only expressions are deprecated

  on main.tf line 5, in locals:
   5:     "${local.dynamic}" = "else"

Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.

Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
```

As suggested, you can suppress the warning by wrapping the dynamic attribute name in parentheses:

```hcl
locals {
  dynamic     = "key_b"
  my_map_json = jsonencode({
    key_a           = "something"
    (local.dynamic) = "else"
  })
  # ...
}
```

I'm marking this issue as closed since there's a fix already merged. This improved warning will be released with Terraform 0.15.0, but the advice still applied to 0.14 and below. Hope this helps!

**Comment by aaronsteers** [comment]
This is great! I wasn't aware of the parenthesis syntax. Thank you!!
